src/AbstractLinesBeforeNamespaceFixer.php

```diff
@@ -34,23 +35,74 @@
      */
     protected function fixLinesBeforeNamespace(Tokens $tokens, $index, $expected)
     {
-        // if we've got a <?php, then subtracted the number of new lines it
-        // contains from the expected number in the following whitespace
-        if (isset($tokens[$index - 2])) {
-            $opening = $tokens[$index - 2];
-            if ($opening->isGivenKind(T_OPEN_TAG)) {
-                $expected -= substr_count($opening->getContent(), "\n");
+        // Let's determine the total numbers of new lines before the namespace
+        // and the opening token
+        $precedingNewlinesTotal = 0;
+        $precedingNewlinesInOpening = 0;
+        $openingToken = null;
+        for ($i = 1; $i <= 2; ++$i) {
+            if (isset($tokens[$index - $i])) {
+                $token = $tokens[$index - $i];
+                if ($token->isGivenKind(T_OPEN_TAG)) {
+                    $openingToken = $token;
+                    $openingTokenIndex = $index - $i;
+                    $precedingNewlinesInOpening = substr_count($token->getContent(), "\n");
+                    $precedingNewlinesTotal += $precedingNewlinesInOpening;
+                    break;
+                }
+                if (false === $token->isGivenKind(T_WHITESPACE)) {
+                    break;
+                }
+                $precedingNewlinesTotal += substr_count($token->getContent(), "\n");
             }
         }
 
+        if ($expected === $precedingNewlinesTotal) {
+            return;
+        }
+
         $previousIndex = $index - 1;
         $previous = $tokens[$previousIndex];
+
+        if (0 === $expected) {
+            // Remove all the previous new lines
+            if ($previous->isWhitespace()) {
+                $tokens->clearAt($previousIndex);
+            }
+            // Remove new lines in opening token
+            if (0 < $precedingNewlinesInOpening) {
+                $tokens[$openingTokenIndex] = new Token(array(T_OPEN_TAG, rtrim($openingToken->getContent()).' '));
+            }
+
+            return;
+        }
+
+        $lineEnding = $this->whitespacesConfig->getLineEnding();
+        if (null !== $openingToken) {
+            $content = rtrim($openingToken->getContent());
+            if (0 === $precedingNewlinesInOpening) {
+                // We have an opening tag without new lines: add a new line there
+                $newContent = $content.$lineEnding;
+                ++$precedingNewlinesInOpening;
+            } else {
+                // Use the configured line ending for the PHP opening tag
+                $newContent = $content.str_repeat($lineEnding, $precedingNewlinesInOpening);
```

Comment:
Not sure you need to use `str_repeat()` as the open tag never contains more than one newline.

Reply:
> Not sure you need to use str_repeat() as the open tag never contains more than one newline.

I agree: I refactored the method to store a boolean value to say if there's a PHP opening tag with a new line or not.